slowmovie.py

```diff
@@ -10,205 +10,233 @@
+
+def generate_frame(in_filename, out_filename, time):
     (
         ffmpeg
         .input(in_filename, ss=time)
-        .filter('scale', width, height, force_original_aspect_ratio=1)
-        .filter('pad', width, height, -1, -1)
-        .output(out_filename, vframes=1)              
+        .filter("scale", "iw*sar", "ih")
+        .filter("scale", width, height, force_original_aspect_ratio=1)
+        .filter("pad", width, height, -1, -1)
+        .output(out_filename, vframes=1)
         .overwrite_output()
         .run(capture_stdout=True, capture_stderr=True)
     )
 
 def check_mp4(value):
-    if not value.endswith('.mp4'):
-        raise argparse.ArgumentTypeError("%s should be an .mp4 file" % value)
+    if not os.path.isfile(value):
+        raise argparse.ArgumentTypeError("File '%s' does not exist" % value)
+    if not supported_filetype(value):
+        raise argparse.ArgumentTypeError("'%s' is not a supported file type" % value)
     return value
 
-# Ensure this is the correct path to your video folder 
-viddir = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'Videos/')
-logdir = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'logs/')
-
-
-parser = argparse.ArgumentParser(description='SlowMovie Settings')
-parser.add_argument('-r', '--random', action='store_true', 
-    help="Random mode: chooses a random frame every refresh")
-parser.add_argument('-f', '--file', type=check_mp4,
-    help="Add a filename to start playing a specific film. Otherwise will pick a random file, and will move to another film randomly afterwards.")
-parser.add_argument('-d', '--delay',  default=120, 
-    help="Delay between screen updates, in seconds")
-parser.add_argument('-i', '--inc',  default=4, 
-    help="Number of frames skipped between screen updates")
-parser.add_argument('-s', '--start',  
-    help="Start at a specific frame")
+def check_dir(value):
+    if os.path.isdir(value):
+        return value
+    else:
+        raise argparse.ArgumentTypeError("Directory '%s' does not exist" % value)
+
+def supported_filetype(file):
+    _, ext = os.path.splitext(file)
+    return ext.lower() in fileTypes
+
+def video_info(file):
+    videoInfo = ffmpeg.probe(file)
+    frameCount = int(videoInfo["streams"][0]["nb_frames"])
+    framerate = videoInfo["streams"][0]["avg_frame_rate"]
+    framerate = float(Fraction(framerate))
+    frametime = 1000 / framerate
+    return frameCount, framerate, frametime
+
+os.chdir(os.path.dirname(os.path.realpath(__file__)))
+
+if "configargparse" in sys.modules:
+    parser = argparse.ArgumentParser(default_config_files=["slowmovie.conf"])
+else:
+    parser = argparse.ArgumentParser()
+parser.add_argument("-D", "--dir", default = "Videos", type = check_dir, help = "Select video directory")
+parser.add_argument("-r", "--random", action = "store_true", help = "Display random frames")
+parser.add_argument("-R", "--random-file", action = "store_true", help = "Play files in random order")
```

Comment:
aren't ```-R``` and ```-f``` kind of mutually exclusive? Should we put them in an argparse group to reflect that?

Reply:
You're right, they are mutually exclusive (-f overrides -R.) That's a good idea.

Reply:
I added it for now, but I'm not sure it's worth it. It might be simpler to explain the order of precedence. Or maybe there's a better way to do all this. Your idea for playback modes could work.

See, what I thought might be a good idea was to write the current settings to the config file. That way the script can be started again with no arguments and continue with the previous settings. But then, the only way to unset some things would be to edit the config file. I might have to ditch that idea if it makes things too complicated.

Reply:
Is there any reason why ```-f``` and ```-D``` need to be mutually exclusive? ```-D``` sets the directory and ```-f``` sets the file to play from that directory. Why shouldn't a user be able to do both of those things? I see in the code, further down, where the directory is automatically set to the directory containing ```-f```  anyway. 

I think your check works to at least make the order of operations somewhat explainable. Everything won't be perfect in this one PR so as long as it's explainable the entire file loading order can be tackled in another run at it.

Reply:
`-f` sets the directory anyway. If a directory is passed with `-D`, and a file in a different directory with `-f`, `-D` is ignored. If `-f` is a file in the same directory, it's redundant. Could make it single argument that can take either.

I was thinking though, if this is gonna be a service, would it make more sense to have only a config file? Then changing settings would be: edit config file, then restart service (`sudo systemctl restart slowmovie`)